### ISSUE TITLE
New parser request: certbot (for ACME / Let's Encrypt)

### ISSUE THREAD

**Issue by g-v-egidy** [kellyjonbrazil/jc#385]
Let's Encrypt is the largest CA. They developed the ACME protocol and the certbot client to implement this protocol. So certbot has become the de-facto standard for getting and renewing certificates. 

Recently [Google began to push](https://www.chromium.org/Home/chromium-security/root-ca-policy/moving-forward-together/) for reducing the maximum issue length for certificates to 90 days, making manually renewing certificates nearly unfeasible. If this gets accepted by the Browser/CA forum, it would make ACME and certbot even more widely used.

Certbot offers the ["certificates" and "show_account" subcommands](https://eff-certbot.readthedocs.io/en/stable/using.html#managing-certificates). Unfortunately they don't return their data in a structured format that can easily be used in scripts and automation. There have been several feature requests and pull requests to add json in the past: https://github.com/certbot/certbot/issues/3909 https://github.com/certbot/certbot/pull/8437 https://github.com/certbot/certbot/pull/8975 . Unfortunately they all have been closed as stale and the maintainers didn't want to invest time into reviewing and integrating them.

It would be nice if jc could fill the gap and help parsing the output.

**Comment by kellyjonbrazil** [comment]
Thanks for the parser suggestion! I like these - I think it makes sense to add them. I’ll look at getting these in the nest release. Funny they haven’t  implemented JSON output since this is not an old tool.

**Comment by g-v-egidy** [comment]
> Thanks for the parser suggestion! I like these - I think it makes sense to add them. I’ll look at getting these in the nest release. 

Thanks.

certbot has a transactional user interface, nothing wrong with that. But when you want to create an idempotent interface instead you have to check the current status and act upon the difference between the desired state. I want to create such an interface for Ansible and jc is already integrated there in a convenient way, so ...

> Funny they haven’t implemented JSON output since this is not an old tool.

Yes, really strange. And they didn't even have to implement that themselves, they got two full implementations as pull requests, they would just have to look through them and merge.

**Comment by kellyjonbrazil** [comment]
Is there any difference in output with verbose flags set? (`-v`, `-vv`, etc.) If so, could you please update this issue with those samples as well?

**Comment by g-v-egidy** [comment]
Good point with the `-v` flags.

certbot certificates doesn't change much with -v flags, but it outputs the data twice:
```
# certbot -vvvv certificates
Root logging level set at -10
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Starting new HTTP connection (1): r3.o.lencr.org:80
http://r3.o.lencr.org:80 "POST / HTTP/1.1" 200 503
OCSP response for certificate /etc/letsencrypt/live/hostname_cert/cert.pem is signed by the certificate's issuer.
OCSP certificate status for /etc/letsencrypt/live/hostname_cert/cert.pem is: OCSPCertStatus.GOOD
Notifying user: Found the following certs:
Found the following certs:
  Certificate Name: example.com
    Serial Number: 3f7axxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    Key Type: RSA
    Domains: example.com www.example.com
    Expiry Date: 2023-05-11 01:33:10+00:00 (VALID: 63 days)
    Certificate Path: /etc/letsencrypt/live/example.com/fullchain.pem
    Private Key Path: /etc/letsencrypt/live/example.com/privkey.pem

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Found the following certs:
  Certificate Name: example.com
    Serial Number: 3f7axxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    Key Type: RSA
    Domains: example.com www.example.com
    Expiry Date: 2023-05-11 01:33:10+00:00 (VALID: 63 days)
    Certificate Path: /etc/letsencrypt/live/example.com/fullchain.pem
    Private Key Path: /etc/letsencrypt/live/example.com/privkey.pem
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
```

`certbot -v show_account` is the same as without the `-v`. 

`certbot -vv show_account` instead becomes completely unparsable as it contains full http protocol output with request and response headers from when they query the account data at Let's Encrypt. I suggest not trying to implement support for it. It doesn't add any info that would be valuable to parse as json in my opinion.

**Comment by kellyjonbrazil** [comment]
I have a simple `certbot` parser complete. It required some small updates to the `jc.utils` module for the timestamps so it's not as easy to test as placing the file in the plugin directory. (I think it may work if you use the `-r` option so the timestamps are not generated)

https://github.com/kellyjonbrazil/jc/blob/dev/jc/parsers/certbot.py

I'll put this in the next release and I can enhance it later, if needed.

**Comment by kellyjonbrazil** [comment]
Released in v1.23.2

**Comment by g-v-egidy** [comment]
I just did some testing with 1.23.2 and it works well.

Thank you very much for implementing this.
